### PR TITLE
Delete service instance and binding resources

### PR DESCRIPTION
**Pull request by lixilin2301**
This PR is created to merge our [fork](https://github.com/chrismathias/terraform-provider-cf) to the origin.
Please refer to issue #36 and pr #37 . Here simply merge the changes from #37 and fix the empty return.

### CODE REVIEW

**Comment by janosbinder** [comment]
LGTM - this is one of the features that we have on our fork, and it would make a nice addition beyond 1.0

**Comment by gberche-orange** [comment]
@janosbinder @lixilin2301 @chrismathias I'm still uncomfortable to add such support in the provider as detailed into https://github.com/mevansam/terraform-provider-cf/issues/36#issuecomment-372292411 

Maybe the rest of the team @jcarrothers-sap @mevansam @psycofdj @ArthurHlt have a different opinion they'd like to share. Alternatively, this can be discussed interactively in one of our weekly status calls.

**Comment by gberche-orange** [comment]
The usual practice for a service instance instanciated by the terraform provider is to have service binding also managed by terraform (and hence no need to force deletion of service bindings unknown to terraform).

However, after discussions with @janosbinder, I now see cases where service plans and service instances could be provisionned by terraform, while unrelated service instance and service bindings be provisioned outside of terraform, and may require a recursive delete.

We should probably consider supporting an additional distinct "purge" flag, to support cases where the associated service broker is unavailable/out of sync, see related https://docs.cloudfoundry.org/services/managing-service-brokers.html#purge-service and https://docs.cloudfoundry.org/services/managing-service-brokers.html#purge-service-instance and matching Cf CLI commands 


```
$ cf help -a 
[...]
   purge-service-offering                 Recursively remove a service and child objects from Cloud Foundry database without making requests to a service broker
   purge-service-instance                 Recursively remove a service instance and child objects from Cloud Foundry database without making requests to a service broker
```

This corresponds to the `purge` flag in http://apidocs.cloudfoundry.org/4.4.0/service_instances/delete_a_service_instance.html

File: website/docs/r/user_provided_service.html.markdown [hunk 1]
```diff
@@ -54,6 +54,7 @@ The following arguments are supported:
 * `credentials_json` - (Optional) Same as `credentials` but in the form of a stringified JSON object. Conflicts with `credentials` and defaults to empty map.
 * `syslog_drain_url` - (Optional) URL to which logs for bound applications will be streamed. Defaults to empty.
 * `route_service_url` - (Optional) URL to which requests for bound routes will be forwarded. Scheme for this URL must be https and defaults to empty
+* `recursive_delete` - (Optional, Bool) Default: `false`. If set `true`, Cloud Foundry will delete service bindings, service keys, and routes associated with the service instance too.
```
Comment (gberche-orange) [inline website/docs/r/user_provided_service.html.markdown]:
```suggestion
```

There is no recursive delete support in the UPS delete api http://apidocs.cloudfoundry.org/4.4.0/user_provided_service_instances/delete_a_particular_user_provided_service_instance.html

File: website/docs/r/service_instance.html.markdown [hunk 2]
```diff
@@ -32,9 +32,10 @@ The following arguments are supported:
 
 * `name` - (Required, String) The name of the Service Instance in Cloud Foundry
 * `service_plan` - (Required, String) The ID of the [service plan](/docs/providers/cloudfoundry/d/service.html)
-* `space` - (Required, String) The ID of the [space](/docs/providers/cloudfoundry/r/space.html) 
+* `space` - (Required, String) The ID of the [space](/docs/providers/cloudfoundry/r/space.html)
 * `json_params` - (Optional, String) Json string of arbitrary parameters. Some services support providing additional configuration parameters within the provision request. By default, no params are provided.
 * `tags` - (Optional, List) List of instance tags. Some services provide a list of tags that Cloud Foundry delivers in [VCAP_SERVICES Env variables](https://docs.cloudfoundry.org/devguide/deploy-apps/environment-variable.html#VCAP-SERVICES). By default, no tags are assigned.
+* `recursive_delete` - (Optional, Bool) Default: `false`. If set `true`, Cloud Foundry will delete service bindings, service keys, and routes associated with the service instance too.
```
Comment (gberche-orange) [inline website/docs/r/service_instance.html.markdown]:
```suggestion
* `recursive_delete` - (Optional, Bool) Default: `false`. If set `true`, Cloud Foundry will delete any service bindings, service keys, and route mappings associated with the service instance. This flag should only be set when such dependent resources were provisioned outside of terraform, and need removal to enable deletion of the associated service instance.
```

File: cloudfoundry/resource_cf_user_provided_service.go [hunk 3]
```diff
@@ -220,8 +225,9 @@ func resourceUserProvidedServiceDelete(d *schema.ResourceData, meta interface{})
 	session.Log.DebugMessage("begin resourceServiceInstanceDelete")
 
 	sm := session.ServiceManager()
+	recursiveDelete := d.Get("recursive_delete").(bool)
 
-	if err = sm.DeleteServiceInstance(d.Id()); err != nil {
+	if err = sm.DeleteServiceInstance(d.Id(), recursiveDelete); err != nil {
```
Comment (gberche-orange) [inline cloudfoundry/resource_cf_user_provided_service.go]: this should be calling sm.DeleteUserProvidedService() instead 

https://github.com/mevansam/terraform-provider-cf/blob/98f04b8016e18301f3fbee1ea5f922e870cb2feb/cloudfoundry/cfapi/service_manager.go#L633-L637

Note that there is no associated recursive delete for the UPS according to http://apidocs.cloudfoundry.org/4.4.0/user_provided_service_instances/delete_a_particular_user_provided_service_instance.html

File: cloudfoundry/resource_cf_user_provided_service.go [hunk 4]
```diff
@@ -66,6 +66,11 @@ func resourceUserProvidedService() *schema.Resource {
 				ConflictsWith:    []string{"credentials"},
 				DiffSuppressFunc: structure.SuppressJsonDiff,
 			},
+			"recursive_delete": &schema.Schema{
```
Comment (gberche-orange) [inline cloudfoundry/resource_cf_user_provided_service.go]: There is no recursive delete support in the UPS delete api http://apidocs.cloudfoundry.org/4.4.0/user_provided_service_instances/delete_a_particular_user_provided_service_instance.html

File: website/docs/r/user_provided_service.html.markdown (hunk 1 above)
Comment (janosbinder) [reply website/docs/r/user_provided_service.html.markdown]: We targetted this API: http://apidocs.cloudfoundry.org/5.1.0/service_instances/delete_a_service_instance.html The latest version still supports the recursive flag.

**Comment by janosbinder** [comment]
@gberche-orange we have migrated to the upstream repository with a blue-green implementation, and since we do not delete environments these days so often, we close this PR.

**Pull request closed without merging**